transport.go

```diff
@@ -1126,6 +1126,33 @@ func (g *connGroup) releaseConn(c *conn) bool {
 	return true
 }
 
+func performTLSHandshake(ctx context.Context, conn *tls.Conn, deadline time.Time) error {
+	// Handshake requires both read and write deadlines to be set or risks indefinite blocking.
+	err := conn.SetDeadline(deadline)
+	if err != nil {
+		return fmt.Errorf("failed to set connection deadline before TLS handshake: %w", err)
+	}
+
+	errch := make(chan error, 1)
+
+	go func() {
+		defer close(errch)
+		errch <- conn.Handshake()
+	}()
+
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
```

Comment:
Should we always wait for the goroutine to terminate before returning?
```suggestion
		err = ctx.Err()
		<-errch
```